### PR TITLE
Minor DIRCON updates

### PR DESCRIPTION
**Pull request by mposa**
1. Set sparsity patern of DIRCON kinematic constraints (slight performance boost).
2. Added option to DIRCON to skip constraints, but still include their associated force variables.

### CODE REVIEW

**Comment by mposa** [comment]
Let me know what you think of the API.

File: systems/trajectory_optimization/dircon_opt_constraints.cc [hunk 1]
```diff
@@ -245,14 +250,47 @@ DirconKinematicConstraint<T>::DirconKinematicConstraint(
...
-  int j = 0;
-  for (int i=0; i < num_kinematic_constraints_; i++) {
-    if (is_constraint_relative_[i]) {
-      relative_map_(i, j) = 1;
-      j++;
+  // ***Set sparsity pattern***
+  std::vector<std::pair<int, int>> sparsity;
+  // Acceleration constraints are dense in decision variables
+  for (int i = 0; i < num_kinematic_constraints_; i++) {
+    for (int j = 0; j < this->num_vars(); j++) {
```
Comment (yminchen) [inline systems/trajectory_optimization/dircon_opt_constraints.cc]: Maybe `for (int j = 0; j < num_states_ + num_kinematic_constraints_; j++)` is slightly faster?

File: systems/trajectory_optimization/dircon_kinematic_data_set.cc [hunk 2]
```diff
@@ -29,11 +31,28 @@ DirconKinematicDataSet<T>::DirconKinematicDataSet(
...
-  cddot_ = VectorX<T>(constraint_count_);
+
+  int total_count = constraint_count_;  // includes skipped constraints
+
+  // We will not include indices givein in skip_constraint_inds
+  constraint_count_ -= skip_constraint_inds.size();
+  constraint_map_ = MatrixXd::Zero(constraint_count_,
+      constraint_count_ + skip_constraint_inds.size());
+  int j = 0;
+  for (int i = 0; i < constraint_count_; i++) {
```
Comment (yminchen) [inline systems/trajectory_optimization/dircon_kinematic_data_set.cc]: Is this supposed to be `for (int i = 0; i < total_count; i++)`?

File: systems/trajectory_optimization/dircon_opt_constraints.cc (hunk 1 above)
Comment (yminchen) [inline systems/trajectory_optimization/dircon_opt_constraints.cc]: Maybe `for (int j = 0; j < num_states_ + num_kinematic_constraints_; j++)` is slightly faster?

File: systems/trajectory_optimization/dircon_opt_constraints.cc [hunk 3]
```diff
@@ -245,14 +250,47 @@ DirconKinematicConstraint<T>::DirconKinematicConstraint(
...
   }
+
+  // Position constraint only depends on q and any offset variables
+  // Set relative map in the same loop
+  if (type == kAll) {
+    relative_map_ = MatrixXd::Zero(num_kinematic_constraints_, n_relative_);
+    int j = 0;
+
+    for (int i = 0; i < num_kinematic_constraints_; i++) {
+      for (int j = 0; j < num_positions_; j++) {
```
Comment (yminchen) [inline systems/trajectory_optimization/dircon_opt_constraints.cc]: I think the current code works, but there are two `j`'s inside and outside the for loop. We can give them different names to improve readability and avoid future bugs.

**Pull request merged**